components/hab/src/command/pkg/sign.rs

```diff
@@ -1,21 +1,35 @@
-use std::path::Path;
+use std::path::{Path,
+                PathBuf};
 
 use crate::{common::ui::{Status,
                          UIWriter,
                          UI},
-            hcore::crypto::{artifact,
-                            SigKeyPair}};
+            hcore::{crypto::{artifact,
+                             SigKeyPair},
+                    package::{Identifiable,
+                              PackageArchive}}};
 
 use crate::error::Result;
 
 pub fn start(ui: &mut UI, origin: &SigKeyPair, src: &Path, dst: &Path) -> Result<()> {
+    // Detect if what we're about to sign is in fact a PackageArchive and
+    // if so, we need to do some basic PackageIdent field validation.
+    if let Ok(mut archive) = PackageArchive::new(PathBuf::from(dst)) {
+        let ident = archive.ident()?;
+        ui.status(Status::Verifying,
+                  format!("{} has valid PackageIdent fields", ident))?;
+        if !ident.valid() {
+            ident.validate_fields()?;
+        }
+    }
```

Comment:
Rather than have this kind of manual verification, it would be good to see what the blast radius of making `PackageIdent::new` return a `Result` instead, and just do all that verification when you create the type. This could also be incorporated into the `FromStr` implementation, as well... basically, any time you have a `PackageIdent`, you _know_ it's valid, because the compiler made it that way.